Sort Transactions on Accounts Page

carkom opened the pull request:
This is a pretty simple and straight forward PR. I've added sorting to the transactions table.

j-f1 commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -1787,6 +1837,120 @@ export let TransactionTable = forwardRef((props, ref) => {
+  let category = props.categoryGroups.flatMap(cat => {
+    return cat.categories;
+  });
+  let fields = [field] + 's';
```
Why are you implicitly converting an array to a string here?

carkom replied:
Do you mean `let fields = [field] + 's';` ?
Sorry, `field ` was never an array. I mistakenly had the brakets in there. I've fixed it.

j-f1 commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -1787,6 +1837,120 @@ export let TransactionTable = forwardRef((props, ref) => {
+    transactions =
+      ascDesc === 'asc'
+        ? transactions
+            .map(e => {
+              return {
+                ...e,
+                test: props[fields].find(f => {
+                  return f.id === e[field];
+                })
+                  ? props[fields].find(f => {
+                      return f.id === e[field];
+                    }).name
+                  : '',
+              };
+            })
```
This bit seems to be pretty repetitive — can you pull it out? Also, you could simplify it a bit by doing `props[fields].find(f => f.id === e[field])?.name ?? ''`

carkom replied:
Thank you, I knew there was a better way to do it. You're a wizard!

j-f1 commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -1787,6 +1837,120 @@ export let TransactionTable = forwardRef((props, ref) => {
+
+  function sortStrings() {
+    transactions =
+      ascDesc === 'asc'
```
Instead of having two different behaviors here for ascending and descending, couldn’t you always pick one way to sort, and then call `reverse` on the array when you’re done if needed?

carkom replied:
Taking me to school 🚌, I love it! This worked great. thanks for the suggestion.

j-f1 commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -1787,6 +1837,120 @@ export let TransactionTable = forwardRef((props, ref) => {
+            .sort((a, b) => a.test > b.test);
+  }
+
+  if (field) {
+    switch (field) {
```
I think the `switch` will hit the `default` case if `field` is not set, right? (Alternatively, you could start `field` off as `'date'` so that there is always a valid sort field).

j-f1 commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -1787,6 +1837,120 @@ export let TransactionTable = forwardRef((props, ref) => {
+      case 'date':
+        ascDesc === 'asc'
+          ? transactions.sort((a, b) => a[field] < b[field])
+          : transactions.sort((a, b) => a[field] > b[field]);
```
All of this is mutating the `transactions` array which is undesirable IMO — especially since it runs on every render, which could potentially be slow if looking at a large list of transactions. Could you move it into a `useMemo` that returns a new array so that it only needs to run if one of the relevant props change? That would also make it possible to factor it out into a custom `useSortedTransactions` hook that could be used in other transaction tables, potentially.

carkom replied:
Makes sense, I'll make these changes.

j-f1 commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -257,14 +266,49 @@ const TransactionHeader = memo(
+          value="Date"
+          width={110}
+          icon={field === 'date' && ascDesc}
+          onClick={e => onSortTable(e)}
```
There isn’t any visible or accessible affordance for this — I think it would be better to show a button on hover (maybe ↕️?) and then have the icon on that button turn into a permanently visible up/down arrow when filtering is active on that column. (to be fair, I haven’t been able to test out the interaction since I can’t get it to work on the deploy preview)

carkom replied:
My thought was that hover pop-ups may be distracting and unneccessary. I opted for the "discoverable features" methodology here. Most users understand the idea of sorting columns by clicking on headers. Plus the default "Date" you suggested shows an up arrow on initial render so it hints people toward the fact that all the columns are sortable.

Thoughts?

carkom replied:
I think I'll wait for you to see it before I make any changes. 

Any idea why the preview isn't building properly?

j-f1 replied:
Preview seems to be fine, the end-to-end tests are failing because GitHub is once again broken 🤷

Tested again, and I like how it works over all! The icon could maybe stand to be a little bit smaller — right now it feels to me like it overpowers the column header text (although then it might get a little bit too thin if it was made smaller).

I like the idea of discoverable features (and think that it would definitely not make sense to show all of the buttons all the time. That said, I don’t think there are really any places in the UI where you are expected to click on a component that shows no signs of being interactive in order to perform an action. Wrapping the icon in a borderless button (like the Import button above it) would make it much clearer that you can click on it. Showing the buttons when hovering a column would also make it possible to know whether or not the column can be sorted without having to click.

carkom replied:
Wrapped it in the borderless button. I like the feel of it better. Thanks for the guidance.

j-f1 commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -257,14 +266,55 @@ const TransactionHeader = memo(
+        />
+        {showAccount && (
+          <Cell
+            value="Account"
+            id="account"
+            width="flex"
+            icon={field === 'account' && ascDesc}
+            onClick={e => onSortTable(e)}
```
This would let you completely avoid the DOM access in the event handler

```suggestion
            value="Account"
            width="flex"
            icon={field === 'account' && ascDesc}
            onClick={() => onSortTable("account")}
```

j-f1 commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -257,14 +266,55 @@ const TransactionHeader = memo(
+            onClick={e => onSortTable(e)}
+          />
+        )}
+        <Cell
+          value="Payment"
+          id="payment"
+          width={90}
+          textAlign="flex"
+          icon={field === 'amount' && ascDesc}
+          onClick={e => onSortTable(e)}
+        />
+        <Cell
+          value="Deposit"
+          id="deposit"
+          width={85}
+          textAlign="flex"
+          onClick={e => onSortTable(e)}
+        />
```
It felt a bit weird that I could sort by payment but not deposit — maybe it would make sense to move the indicator to the Deposit icon so the two columns feel more joined than separated by the icon?

j-f1 commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -1787,6 +1843,77 @@ export let TransactionTable = forwardRef((props, ref) => {
+          fieldName: category.find(f => f.id === e[field])?.name ?? '',
+        };
+      })
+      .sort((a, b) => a.fieldName > b.fieldName);
```
Sorters are supposed to return a number (`-1` for less than, `0` for equal, `1` for greater than) rather than a boolean. This might be causing the instability, but I’m not sure. For strings, I would recommend using `a.localeCompare(b)`. For numbers I think the convention tends to be `a - b`.

j-f1 reviewed (commented):
Good work so far. I’m noticing an issue with the demo budget when I sort by date (reversed). My suggestion below might help but you might also need to change the sorter to produce an unambiguous result by falling back to sorting by UUID as a fallback if the two transactions match?

https://github.com/actualbudget/actual/assets/25517624/3f1ef7a0-0726-4711-ac7e-dd18aa631b4f

carkom commented:
Something from a recent merge is causing my page to freak out...bouncing between category and amount as if they are being clicked on but they aren't. @j-f1 can you help me find the bug?

j-f1 commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -257,14 +266,49 @@ const TransactionHeader = memo(
+          value="Date"
+          width={110}
+          icon={field === 'date' && ascDesc}
+          onClick={onSortTable('date')}
```
This will call `onSortTable('date')` on every render, you’ll need to wrap it in a function in order for it to run on click. (same for all of the other ones)

```suggestion
          onClick={() => onSortTable('date')}
```

carkom replied:
🤦‍♂️ . Thanks.

carkom commented:
> I’m noticing an issue with the demo budget when I sort by date (reversed). 

I think the usememo addition fixed this. Can you double check it?

j-f1 commented on packages/desktop-client/src/components/table.tsx:
```diff
@@ -261,7 +265,21 @@ export function Cell({
           >
-            {formatter ? formatter(value) : value}
+            {icon ? (
+              <Button bare>
```
Very small UI nits:

* the ones with a button are a slightly different appearance than the ones without (they appear to be a darker color & have font weight 400 instead of 300), it would be nice to keep them looking the same as now
* it feels a little weird to me that clicking on “Payment” works but doesn’t add an icon there… it would be nice to have the button background extend to cover both the Payment and Deposit amounts (or something?). Sounds challenging, feel free to punt this detail to an issue/future PR if you can’t come up with a way to do something similar
* The natural padding of the `Button` brings the start of the heading title slightly to the right of the text inside the table. The plain text labels look to be about 1px too far to the left on `master`, so it would be nice to tighten those up along with adding a negative left margin for the button to make sure all columns line up well.

carkom replied:
- Changed button color and weight to match existing table header so they don't look different from `master`.
- I gave payment and deposit their own icons. They effectively do the same thing but have different default behavior. This does feel like a better UI, thanks!
- negative left margin cuts off the button. Not sure the best option as they don't line up in master...(have a look at balance column in master)

j-f1 replied:
> negative left margin cuts off the button

It looks like one of the parent elements (the direct parent of the `<button>` in the HTML) has `overflow: hidden` which is causing this. Might be best to switch that to `overflow: visible`, at least in the X direction? Although I’m not sure why that is there, maybe there’s a good reason

> Not sure the best option as they don't line up in master

I think payment/deposit are right-aligned, the others are left-aligned?

j-f1 reviewed (commented):
Confirmed the sorting is now stable.

carkom commented:
I also boosted resilience by adding "date" as a secondary sort field as well as UUID as a 3rd. This way the sort is always the same 100% of the time. (For the date field I defaulted the secondary sort to "Payee".

carkom commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -1847,12 +1898,96 @@ export let TransactionTable = forwardRef((props, ref) => {
+        };
+      })
+      //if 2 transactions share a date and fieldname then sort by UUID
+      .sort((a, b) => a.id.localeCompare(b.id))
+      //secondary: date
+      .sort((a, b) => a.date.localeCompare(b.date))
```
Changed logic

carkom commented on packages/desktop-client/src/components/accounts/TransactionsTable.js:
```diff
@@ -1847,12 +1898,96 @@ export let TransactionTable = forwardRef((props, ref) => {
+      //primary field
+      .sort((a, b) => a.fieldName.localeCompare(b.fieldName)).sort((a, b) => a.id - b.id);
+
+    return ascDesc === 'desc' ? sortDesc : sortDesc.reverse();
```
Only problem is that they don't all default to sorting the same way. Default sort for Date is most recent at the top and default for any words is "A" at the top. I've changed the functionality a bit so this may be outdated.

carkom commented:
I've added a previousSort into the sorting so that the sort uses most recent action as first sort column and last action as next sort column.

I've added a UUID default sort in case current and prev sort are exactly the same. This allows the sorting to be exactly the same everytime you take the same actions in the same sequence.

carkom commented:
I think the only way to truly fix the padding/alignment issues is if I can read the scrollbar status. @j-f1 is there a way to get the state of the scroll bar (hidden/visible). I've been searching for an hour and just can't find it.

j-f1 commented:
I’m not quite sure what you mean by “scrollbar status.” The scrollbar is provided by the OS and we don’t have any control over it really. What are you trying to do?

carkom commented:
As an example from Rule modal. Goal is to leave space for the columns when scrollbar is hidden and to remove that space when it is present. This will allow the transaction columns to stay even with the headers under all circumstances.

![image](https://github.com/actualbudget/actual/assets/55785687/36752c61-ab6e-4f26-91f6-af389bf92bf0)

j-f1 commented:
Oh, I see… the best option I can think of would be to compare the width of the element containing the scrollable element with the width of an element inside the scrollable element, e.g.:

```jsx
// very rough sketch:
<div id="outside">
  <div style={{ overflow: 'auto' }}>
    <div id="inside">
    </div>
  </div>
</div>

const scrollbarWidth = $('#outside').clientWidth - $('#inside').clientWidth;
```

This would then need to be re-run whenever the window resizes / the height of the elements above/below the scrollable area change / the scrollable content changes.

carkom commented:
Ahh okay. Was hoping for some sort of `showScrollbar:` type of function. Showing `true` when it's there and `false` when it's not.

j-f1 commented:
I’m not saying there isn’t such a thing but I’m guessing you would have found it already if it did exist :)

carkom commented:
I've found a "scrollContainer" `useRef` item in the transactions table function that housed the table width and the parent component width. Slapped that into a useState and it's now working as intended. I'm happy with all the column alignments now.

I don't have any other work to finish up on this one. Let me know if any other bugs are spotted, otherwise I'd say ready to merge.

MatissJanis commented:
👋 This does not seem to be working for bigger budgets. So lets say I have 10'000 transactions. Only 100 are rendered on the table. And only the 100 will be sorted.

If I scroll down and load the new page (so 200 in total when I reach page 2) - it will sort the 200 transactions only.

carkom commented:
> 👋 This does not seem to be working for bigger budgets. So lets say I have 10'000 transactions. Only 100 are rendered on the table. And only the 100 will be sorted.
> 
> If I scroll down and load the new page (so 200 in total when I reach page 2) - it will sort the 200 transactions only.

Good catch. Do you have any ideas on this? I've tried everything I can think of to get the transactions sorted before it renders the first 100 rows but nothing works. 

Any guidance?

j-f1 commented:
I think you will need to pass the sort information back up to the component that renders the transaction table, then add a `orderBy` call to the query builder that fetches the list of transactions.

carkom commented:
Hey all, this is ready for review. I've fixed the sort to use order by at the query level.

I'm not sure what the E2E test error is. Something about netlify creating rules but when I try it manually it works fine. 🤷‍♂️ 

@j-f1 @MatissJanis

Kidglove57 commented:
I confirm this is working for me - even with a transaction register that goes back to 2019.

It has the very happy knock-on benefit of Filters and Saved Filters also being sortable. Something that I have wanted for a long time.

Thank you so much for working on this! My personal must-have list for Actual is nearing completion, thanks to efforts like yours.

j-f1 commented:
I’m seeing a crash when opening the payees page because the `getScrollWidth` prop is not being passed. Is there a way to make this prop optional, or ensure that all call sites use it?

j-f1 commented:
Otherwise this looks great!

carkom commented:
> I’m seeing a crash when opening the payees page because the `getScrollWidth` prop is not being passed. Is there a way to make this prop optional, or ensure that all call sites use it?

Try that.

j-f1 commented:
Thanks! That works now. Comparing with `master`, it looks like all columns have gained an additional 5px of left padding. Is there a way to remove that? It’s a pretty small amount but it does ever so slightly decrease the amount of information that can be displayed at a given screen width.

carkom commented:
> Thanks! That works now. Comparing with `master`, it looks like all columns have gained an additional 5px of left padding. Is there a way to remove that? It’s a pretty small amount but it does ever so slightly decrease the amount of information that can be displayed at a given screen width.

This goes back to a previous discussion we had about the button padding. Without the column padding the headers are 5px to the right of the column data because they are wrapped in a button.
https://github.com/actualbudget/actual/pull/1232#discussion_r1251073037

j-f1 commented on packages/desktop-client/src/components/table.tsx:
```diff
@@ -175,7 +178,35 @@ export function UnexposedCellContent({
         textOverflow: 'ellipsis',
       }}
     >
```
The reason adding a negative left margin to the button results in truncation is that this element wraps the `<Button>`. I think the better approach here would be to use the `unexposedContent` prop to inject the `<Button>` directly without this wrapping `<Text>` (maybe using a component that wraps `<Cell>` so that logic doesn’t need to be duplicated too much?) That would allow you to remove the extra padding inside the table. (Also, adding the `whiteSpace: 'nowrap', overflow: 'hidden', textOverflow: 'ellipsis'` combo to the `<Button>` content would help the text truncate more gracefully on narrow window widths)

carkom replied:
Thanks for the suggestions. I think we got there. Let me know!

carkom commented:
Hey @j-f1, I think this one is ready.

j-f1 commented:
It looks like there are a couple of changes to the transaction table layout in this PR:

- The “cleared” checkboxes are now smaller
- The money amounts are now left- instead of right-aligned.

The former is just cosmetic (but should be fixed IMO as the icon is very small now). The second one is causing the E2E test failure because of how the test works.

carkom commented:
>     * The “cleared” checkboxes are now smaller
Yes I did this becuase the alignment was not consistent. By using a fixed {21} in the header and an "auto" in the table it was causing the table columns to not line up properly with the headers. My proposed fix was to use fixed widths for both columns and headers to help make things consistent. Easiest fix is to make the fixed width larger {30} seems to keep the check boxes the same size as master while keeping the alignments correct. Thoughts?

>     * The money amounts are now left- instead of right-aligned.
Again, this became an alignment issue. The alignment of the columns in master is a bit wonky already. By creating a header button with a hidden icon that displays once clicked (sorted) the alignment of the numbers was way off and it made the table look aweful. Is there anyway to change the logic for the test so that we can keep the alignment of the columns in line with the headers? I think a change to the master is needed. If you look at how it is now the `Balance` header looks really strange due to it's mis-alignment (the last "e" in the word is completly hanging over into the next column) with it's data column underneath. My PR proposes consistency and a tidier cleaner look.

j-f1 commented:
> Easiest fix is to make the fixed width larger {30} seems to keep the check boxes the same size as master while keeping the alignments correct. Thoughts?

That seems reasonable!

> Again, this became an alignment issue. The alignment of the columns in master is a bit wonky already. By creating a header button with a hidden icon that displays once clicked (sorted) the alignment of the numbers was way off and it made the table look aweful. Is there anyway to change the logic for the test so that we can keep the alignment of the columns in line with the headers? I think a change to the master is needed. If you look at how it is now the Balance header looks really strange due to it's mis-alignment (the last "e" in the word is completly hanging over into the next column) with it's data column underneath. My PR proposes consistency and a tidier cleaner look.

This is a tricky problem…

The currency amounts definitely need to be right-aligned so that the `.00` part is lined up and it’s easy to scan the amounts at a glance (virtually all accounting software does this, we should follow in their footsteps). One option could be to move the up/down arrow to the left side of the label for the amount columns (not sure if that would look bad?)


That said, I also just noticed an issue with sorting: the balance column now makes a lot less sense since the table isn’t always chronologically ordered anymore. This means the numbers can end up looking incorrect (e.g. if you sort by amount the account goes way into the red before showing all the deposits). Not sure what the best way to deal with this would be.

Kidglove57 commented:
As regards the balance column - I just opened up another piece of mature personal finance software that I have installed - Moneywiz. The way they handle this is interesting but logical:
1) When sorting by date, each transaction retains its original balance figure. I believe that Actuall should do the same (it does not currently in this PR - the balance is meaningless).
2) When sorting by **any** coumn, other than date, the balance column is hidden because it is no longer relevant. Again a sensible and pragmatic approach.
I would be very happy for Actual to adopt this approach.

carkom commented:
@j-f1 , understood on the debit/credit/balance alignment. I think I've got it sorted. Have a look.

Waiting to hear back on how we want to handle the balance column issue. It seems like hiding it when any column is sorted is the best UI from my perspective. Happy to hear others views as well. 

@Kidglove57, thanks for the reference points. What's your preference on best UX?

Kidglove57 commented:
@carcom thank you! Yes I'm personally happiest with the balance column being hidden on sorting, as it becomes meaningless.

j-f1 commented:
> understood on the debit/credit/balance alignment. I think I've got it sorted. Have a look.

That feels a lot better!

I also feel like hiding the balance column is reasonable if the sort is changed. Maybe that could be combined with updating the “Show Running Balance” menu item to also reset the sort before showing the column so that the user can always get the column back without struggling with resetting the sort manually?

carkom commented:
Okay, sorting now removes balance column. Show Balance column menu item now also clears filters, sorting and searches. 

I aslo added a tooltip for the `Show Balance` menu item since it is such a meaningful switch and the user could lose all work they've done on the current view. If you think it's out of place, let me know it's easy to remove it.

Thanks for your suggestions, this is feeling really nicely polished.

Kidglove57 commented:
Thank you very much - this very much feels like a move in the the right direction to me as a user. A few thoughts from experimenting with it:
1) When using the existing filter tool (in 23.7.2) selecting "clear all conditions" *automatically* returns the register to date sorting and shows the balance column again, without this column needing to be reselected.
2) However,  this very helpful behaviour is lost when clearing a filter in the current PR. No doubt as a corollary to the proposed balance column fix for sorting.
3) It would be great (if possible) to retain this existing filter experience under the current PR.

Therefore, my ideal experience when sorting would be similar to the current experience when filtering. To achieve this, I would personally like to:
a) Always **retain** the balance column when changing the sorting by date from descending to ascending. The balance shown would remain the one pertaining to the individual transaction rows before the sorting.
b) When sorting by any other column except date (say by category) the user will normally always finish by re-sorting by date/descending. Whereupon my wish would be for the balance column to automatically reappear without needing to be reselected.

From a coding perspective is this possible, please? I think it could end up annoying users having to specifically choose to display the balance column again after each sort is finished and even after they have returned to date order. I am hoping that retaining the balance column view, but for date sorting only, would get over this?

carkom commented:
>     2. This very helpful behaviour is retained when clearing a filter in the current PR **unless** a sort has also been applied to the filtered view. This leads me to ask if the logic used in the filter tool (before sorting) could be applied to sorting generally?
I think you are making an assumption on how the feature should work based on how you would like it to work. There's also use cases where someone might want to clear filters and retain the current sort. I don't think we should build a "clear sort" element into the clear filters option, they should remain separate IMO.

> BUT the user will normally always finish by re-sorting by date/descending. 
This is another assumption based on how you use the app. I don't think we can say how anyone "normally always" does anything in the app. So to base our UX on an assumption is a bad idea.

> Whereupon my wish would be for the balance column to automatically reappear without needing to be reselected. c) However, I am not sure how to create a cohesive experience regarding the balance column if a filter has been applied **and** then a sort as well. Which I guess is why you have made the latest changes in the way you have?
Correct, filters and sorting are treated separately. I think this is the only logical way to treat these features for the most sensible use cases for **ALL** users.

> From a coding perspective is it in any way possible, please, to always return to balance column view when reselecting date sorting, even if a filtered view has been sorted? I think it could end up annoying users having to specifically choose to display the balance column again after each sort is finished and even after they have returned to date order. 
Keep in mind the the balance column is non-sensical with any filters applied. As such, you're suggesting that anytime date sorting is selected we clear all filters and restore the balance column. How would someone sort by date with a filter applied? With the above suggestion it wouldn't be possible, or at least you'd have to do it in a specific order to get it to work. This would be a really consfusing action from the app IMO. Alternatively, as it works right now, you can restore the balance column by choosing to restore the balance column.

Kidglove57 commented:
Just to be clear, I will use the sort feature regardless of the final decision on the UI. I want it badly and I am grateful to you for making that possible.

This is just one user's feedback and it would of course be great to have that of others. 

For me, as one user, there was simply a lack of cohesion in the user experience:
A) on sorting, compared to other apps I have access to, which restored the balance column without any user intervention, once the date column was used again for sorting. Which surely would be almost everyone's default view for transaction entry?
B) Between the filter UX and the sorting UX.

But as I said, I will leave it for others to comment. I remain in your debt for taking this on.

carkom commented:
@Kidglove57, I think I found a good solution. I copied the logic from filters and searches. Sorting removes the balance column but remembers your last show/hide choice of the balance column option. `Sort date desc` acts as a "delete all sorts" button and allows the balance to show again. However, filter and search also need to be empty for this to work, I still think we need to keep them as separate elements that are controlled independently. Let me know if this is closer to what you meant.

Kidglove57 commented:
@carkom Thank you very much for working out a solution - much appreciated. 

1) My concern was to provide a cohesive experience for those coming from the current fixed view (Date:Desc) experience. Your update achieves this perfectly since, as you say, clicking Date:Desc restores the expected view after sorting by other columns. Or, of course, clicking "Remove all Sorting" would achieve the same thing once users have discovered that option.
2) I have run through each column sort using my live budget file which includes 3 years of data and over 20 accounts. All sorts worked for me as expected. The only column which looked a bit odd was Payees. The "real" payees are sorted perfectly but the transfers between accounts in the payee column did not list in any discernable order. I can live with that. Here is an example from the Payee column with Payee:Ascending selected:
<img width="215" alt="Payee Ascending@2x" src="https://github.com/actualbudget/actual/assets/93882449/ca8171fc-6664-4142-9ca8-7e3f58053f2c">
3) Is there a reason why the "Remove All Sorting" option appears in the individual accounts view but not in All Accounts or For Budget?
4) The interaction between Filters and Sorting is I agree problematic since they operate independently. For example:
-  Creating a filter and then sorting by, say, payee works well. If I then click "Remove all Sorting" the filter goes haywire. However, if I then click Date:Desc the filter is restored. Or I can just click Date:Desc and not use the Remove all Sorting at all.

Kidglove57 commented:
Thanks so much @carkom. And for the clear explanations.

I have tested again with my complex file and it now all works beautifully for me. I'm loving this extra functionality.

I agree about the separate PR for transfer account name issue.

j-f1 commented on packages/desktop-client/src/components/accounts/Account.js:
```diff
@@ -523,11 +551,27 @@ class AccountInternal extends PureComponent {
           this.calculateBalances();
         }
         break;
+      case 'remove-sorting':
+        let filters = this.state.filters;
+        this.setState({ sort: [] });
+        filters.length > 0
+          ? this.applyFilters([...filters])
+          : this.fetchTransactions();
+        this.state.search !== '' && this.onSearch(this.state.search);
+        break;
```
Since this isn’t an expression can you expand it to an `if` statement? I don’t think we really use the ternary pattern like this in the rest of the codebase. Also, it’s best to use an explicit block to avoid issues if a different `case` wants to declare a `filters` variable:
```suggestion
      case 'remove-sorting': {
        let filters = this.state.filters;
        this.setState({ sort: [] });
        if (filters.length > 0) {
          this.applyFilters([...filters])
        } else {
          this.fetchTransactions();
        }
        this.state.search !== '' && this.onSearch(this.state.search);
        break;
      }
```

j-f1 commented on packages/desktop-client/src/components/table.tsx:
```diff
@@ -1014,6 +1020,15 @@ export const Table = forwardRef<TableHandleRef, TableProps>(
       let editing = editingId === item.id;
       let selected = isSelected && isSelected(item.id);
 
+      scrollContainer.current &&
+        getScrollWidth &&
+        getScrollWidth(
+          scrollContainer.current.offsetParent
+            ? scrollContainer.current.offsetParent.clientWidth
+            : 0,
+          scrollContainer.current ? scrollContainer.current.clientWidth : 0,
+        );
```
Since `getScrollWidth` doesn’t return anything, `get` doesn’t feel like the right name to me. How about something like `saveScrollWidth` or `exportScrollWidth` or `setScrollWidth` etc?

also nit: prefer using an if statement here (will try to find an ESLint rule for this!)

```suggestion
      if (scrollContainer.current && getScrollWidth) {
        getScrollWidth(
          scrollContainer.current.offsetParent
            ? scrollContainer.current.offsetParent.clientWidth
            : 0,
          scrollContainer.current ? scrollContainer.current.clientWidth : 0,
        );
      }
```

j-f1 replied:
Opened #1355 for the second part.

j-f1 reviewed (commented):
Looks mostly good to me, just a few minor nits

j-f1 reviewed (approved):
LGTM but I want @MatissJanis to take a look too.

MatissJanis reviewed (approved):
Thanks! This is a great addition to Actual!
